### PR TITLE
Adds a new marker interface HttpRequestTenantResolver

### PR DESCRIPTION
**Pull request by sdelamo**
- cookie
- subdomain
- principal 
- http header 
- session

resolvers implement this marker. 

This allows to resolve the current tenant without using `ServerRequestContext.currentRequest();`

### CODE REVIEW

File: multitenancy/src/main/java/io/micronaut/multitenancy/tenantresolver/HttpRequestTenantResolver.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * @since 2.1.2
+ */
+public interface HttpRequestTenantResolver {
+    /**
+     * Resolves the current tenant identifier.
+     * @param request The HTTP request
+     * @return The tenant identifier
+     * @throws TenantNotFoundException if tenant not found
+     */
+    Serializable resolveTenantIdentifier(@NonNull @NotNull HttpRequest<?> request) throws TenantNotFoundException;
```
Comment (graemerocher) [inline multitenancy/src/main/java/io/micronaut/multitenancy/tenantresolver/HttpRequestTenantResolver.java:40]: Needs to expression whether the return type is nullable or not

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: multitenancy/src/main/java/io/micronaut/multitenancy/tenantresolver/HttpRequestTenantResolver.java (hunk 1 above)
Comment (sdelamo) [reply multitenancy/src/main/java/io/micronaut/multitenancy/tenantresolver/HttpRequestTenantResolver.java:40]: I've added the `@NonNull` annotation to both `TenantResolver::resolveIdentifier`  and `HttpRequestTenantResolver::resolveIdentifier(HttpRequest)` methods.

**Comment by sdelamo** [comment]
Moreover, I've annotated  both `TenantResolver`  and `HttpRequestTenantResolver` with `@FunctionalInterface`

**Comment by jameskleeh** [comment]
@sdelamo This should target 2.2

**Pull request merged**